dd-sdk-android-ndk/src/main/cpp/utils/signal-monitor.c

```diff
@@ -0,0 +1,193 @@
+                                &global_previous_sigaction[i]);
+        if (success != 0) {
+            pthread_mutex_unlock(&handlerMutex);
+            return false;
+        }
+    }
+
+    pthread_mutex_unlock(&handlerMutex);
+
+    return true;
+}
+
+void uninstall_signal_handlers() {
+    if (global_env == NULL)
+        return;
+
+    pthread_mutex_lock(&handlerMutex);
+
+    if (global_env == NULL) {
+        return;
+    }
+    size_t signals_array_size = handled_signals_size();
+    for (int i = 0; i < signals_array_size; i++) {
+        const int signal = handled_signals[i].signal_value;
+        sigaction(signal, &global_previous_sigaction[i], 0);
+    }
+    global_env = NULL;
+
+    pthread_mutex_unlock(&handlerMutex);
+}
+
+bool configure_signal_stack() {
+    static size_t stackSize = SIGSTKSZ * 2;
+    if ((signal_stack.ss_sp = calloc(1, stackSize)) == NULL) {
+        return false;
+    }
+    signal_stack.ss_size = stackSize;
+    signal_stack.ss_flags = 0;
+    if (sigaltstack(&signal_stack, 0) < 0) {
+        return false;
+    }
+    return true;
+}
+
+void invoke_previous_handler(int signum, siginfo_t *info,
+                             void *user_context) {
+
+    pthread_mutex_lock(&handlerMutex);
+
+    size_t signals_array_size = handled_signals_size();
+    for (int i = 0; i < signals_array_size; ++i) {
+        const int signal = handled_signals[i].signal_value;
+        if (signal == signum) {
+            struct sigaction previous = global_previous_sigaction[i];
+            // From sigaction(2):
+            //  > If act is non-zero, it specifies an action (SIG_DFL, SIG_IGN, or a
+            //  handler routine)
+            if (previous.sa_flags & SA_SIGINFO) {
+                // This handler can handle signal number, info, and user context
+                // (POSIX). From sigaction(2): > If this bit is set, the handler
+                // function is assumed to be pointed to by the sa_sigaction member of
+                // struct sigaction and should match the proto- type shown above or as
+                // below in EXAMPLES.  This bit should not be set when assigning SIG_DFL
+                // or SIG_IGN.
+                previous.sa_sigaction(signum, info, user_context);
+            } else if (previous.sa_handler == SIG_DFL) {
+                // raise to trigger the default handler. It cannot be called directly.
+                raise(signum);
+
+            } else if (previous.sa_handler != SIG_IGN) {
+                // This handler can only handle to signal number (ANSI C)
+                void (*previous_handler)(int) = previous.sa_handler;
+
+                previous_handler(signum);
+            }
+        }
+    }
+
+    pthread_mutex_unlock(&handlerMutex);
+}
+
+void handle_signal(int signum, siginfo_t *info,
+                   void *user_context) {
+    pthread_mutex_lock(&handlerMutex);
+    if (global_env == NULL) {
+        return;
```

Comment:
You need to unlock before returning but in that case I would probably write something like:
```
int is_initialized = 0;
pthread_mutex_lock(&handlerMutex);
is_initialized = (global_env == NULL);
pthread_mutex_unlock(&handlerMutex);
if (is_initialized) return;
```